### PR TITLE
*: fix bug that UnionScan can't keep order caused wrong result

### LINKED ISSUE #33175: the results is wrong because of bug in bigint(45) handle column comparing
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
use test;
source [test.txt](https://github.com/pingcap/tidb/files/8280923/test.txt)
alter table PK_SIGNED_10093 cache;

# you can execute many times to reproduce
select col1, col2 from PK_SIGNED_10093 where col1 > (select max(col1) from PK_SIGNED_10093 t1 where t1.col1 > 0);
```
### 2. What did you expect to see? (Required)
```
MySQL [test]> select col1, col2 from PK_SIGNED_10093 where col1 > (select max(col1) from PK_SIGNED_10093 t1 where t1.col1 > 0);
Empty set (0.00 sec)
```

### 3. What did you see instead (Required)
```
MySQL [test]> select col1, col2 from PK_SIGNED_10093 where col1 > (select max(col1) from PK_SIGNED_10093 t1 where t1.col1 > 0);
+----------------------+--------------------------------------------------------------+
.....
+----------------------+--------------------------------------------------------------+
50 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```
Release Version: v6.0.0-alpha-122-g46c43fe
Edition: Community
Git Commit Hash: 46c43febcac0e24bc9f0872e93cf6125db69d93d
Git Branch: master
UTC Build Time: 2022-03-17 01:02:30
GoVersion: go1.17.6
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #33175 

Problem Summary:

### What is changed and how it works?

Actually, there are two bugs that cause the UnionScan fail to keep order, 
and since it can't keep order, the following plan return wrong result.
`Limit1->UnionScan->XXX` doesn't return the `max(id)`

```
mysql> explain select max(id) from t;
+------------------------------+---------+-----------+---------------+-------------------------------------+
| id                           | estRows | task      | access object | operator info                       |
+------------------------------+---------+-----------+---------------+-------------------------------------+
| StreamAgg_10                 | 1.00    | root      |               | funcs:max([test.t.id](http://test.t.id/))->Column#3      |
| └─Limit_17                   | 1.00    | root      |               | offset:0, count:1                   |
|   └─UnionScan_21             | 1.00    | root      |               |                                     |
|     └─TableReader_23         | 1.00    | root      |               | data:TableFullScan_22               |
|       └─TableFullScan_22     | 1.00    | cop[tikv] | table:t       | keep order:true, desc, stats:pseudo |
+------------------------------+---------+-----------+---------------+-------------------------------------+
5 rows in set (0.00 sec)
```

The first bug is the datum comparing, the data type is KindUint64, but the old code treat it as `GetInt64()`,
10353107668348738101 become a negative number and 10353107668348738101 < 33

The second bug is the handling of kv range, in TiKV, the range is singed int
Here the data type is uint, so correct range handling is:

```
handling range: 7480000000000000415f720000000000000000 7480000000000000415f728000000000000000
handling range: 7480000000000000415f728000000000000000 7480000000000000415f72ffffffffffffffff00
add Row is nil, get snapshot row... [KindUint64 10353107668348738101]
add Row is nil, get snapshot row... [KindUint64 9734095886065816707]
add Row is nil, get snapshot row... [KindUint64 0]
```

the wrong range handling is:

```
handling range: 7480000000000000415f728000000000000000 7480000000000000415f72ffffffffffffffff00
in iterator ... add row == [KindUint64 0]
handling range: 7480000000000000415f720000000000000000 7480000000000000415f728000000000000000
in iterator ... add row == [KindUint64 9734095886065816707]
in iterator ... add row == [KindUint64 10353107668348738101]
```

i.e.  the kv range order should be [-max int64, 0]  [0, max int64]

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix bug that 'select max(PK) ...' query return wrong result, the trigger condition is that UnionScan is used and the type of `PK` is uint64
```

### CODE REVIEW

File: executor/table_reader.go [hunk 1]
```diff
@@ -181,18 +181,18 @@ func (e *TableReaderExecutor) Open(ctx context.Context) error {
...
-		e.kvRanges = append(e.kvRanges, kvReq.KeyRanges...)
 		if len(secondPartRanges) != 0 {
-			kvReq, err = e.buildKVReq(ctx, secondPartRanges)
+			kvReq, err := e.buildKVReq(ctx, secondPartRanges)
 			if err != nil {
 				return err
 			}
 			e.kvRanges = append(e.kvRanges, kvReq.KeyRanges...)
 		}
+		kvReq, err := e.buildKVReq(ctx, firstPartRanges)
```
Comment (winoros) [inline executor/table_reader.go:191]: Add a comment here to tell why we put it after the `second part`.

**Comment by jackysp** [comment]
When was this introduced? @tiancaiamao

**Comment by tiancaiamao** [comment]
> When was this introduced? @tiancaiamao

I believe it's been there for quite a long time...
Maybe v3.0 or even v2.1 ... we already have it @jackysp

File: executor/mem_reader.go [hunk 2]
```diff
@@ -119,6 +119,14 @@ func (m *memIndexReader) getMemRows(ctx context.Context) ([][]types.Datum, error
 	}
 
 	mutableRow := chunk.MutRowFromTypes(m.retFieldTypes)
+
+	// TODO: `IterReverse` is not used... to get the same effect, reverse the kv ranges first,
+	// Then reverse the whole added rows.
+	// [99, 100] [44, 45] [1, 3] => [1, 3] [44, 45] [99, 100] => [100, 99] [45, 44] [3 1]
+	if m.desc {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/mem_reader.go]: Update, PTAL @lcwangchao

File: executor/mem_reader.go [hunk 3]
```diff
@@ -239,6 +239,7 @@ func (m *memTableReader) getMemRows(ctx context.Context) ([][]types.Datum, error
 		defer span1.Finish()
 		opentracing.ContextWithSpan(ctx, span1)
 	}
+
```
Comment (lcwangchao) [inline executor/mem_reader.go:242]:
```suggestion
```

**Review by lcwangchao — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**